File: sdk/python/tonebridge/services/auto_transform.py
"""
Auto-Transform Service
Handles automatic transformation rules and configurations
"""

from typing import Optional, Dict, Any, List, Union
from ..types import (
    AutoTransformConfig,
    AutoTransformRule,
    MessageContext,
    TransformationResult,
    TransformationType,
    TriggerType,
)
from ..constants import API_ENDPOINTS
from ..exceptions import ValidationError


class AutoTransformService:
    """Service for auto-transform operations"""
    
    def __init__(self, client):
        """
        Initialize auto-transform service
        
        Args:
            client: ToneBridgeClient instance
        """
        self.client = client
        self.base_endpoint = "/auto-transform"
    
    def get_config(self, tenant_id: Optional[str] = None) -> Dict[str, Any]:
        """
        Get auto-transform configuration
        
        Args:
            tenant_id: Tenant ID
            
        Returns:
            Configuration dictionary
        """
        endpoint = f"{self.base_endpoint}/config"
        if tenant_id:
            endpoint = f"{endpoint}/{tenant_id}"
        
        return self.client.request("GET", endpoint)
    
    def update_config(
        self,
        config: Union[AutoTransformConfig, Dict[str, Any]],
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Update auto-transform configuration
        
        Args:
            config: Configuration to update
            tenant_id: Tenant ID
            
        Returns:
            Updated configuration
        """
        endpoint = f"{self.base_endpoint}/config"
        if tenant_id:
            endpoint = f"{endpoint}/{tenant_id}"
        
        config_data = config.__dict__ if hasattr(config, '__dict__') else config
        return self.client.request("PUT", endpoint, json=config_data)
    
    def enable(self, tenant_id: Optional[str] = None) -> None:
        """
        Enable auto-transform
        
        Args:
            tenant_id: Tenant ID
        """
        self.update_config({"enabled": True}, tenant_id)
    
    def disable(self, tenant_id: Optional[str] = None) -> None:
        """
        Disable auto-transform
        
        Args:
            tenant_id: Tenant ID
        """
        self.update_config({"enabled": False}, tenant_id)
    
    def evaluate(
        self,
        context: Union[MessageContext, Dict[str, Any]],
    ) -> Dict[str, Any]:
        """
        Evaluate if message should be auto-transformed
        
        Args:
            context: Message context
            
        Returns:
            Transformation evaluation result
        """
        context_data = context.__dict__ if hasattr(context, '__dict__') else context
        self._validate_message_context(context_data)
        
        return self.client.request(
            "POST",
            f"{self.base_endpoint}/evaluate",
            json=context_data
        )
    
    def transform(
        self,
        context: Union[MessageContext, Dict[str, Any]],
        transformation: Union[TransformationResult, Dict[str, Any]],
    ) -> Dict[str, Any]:
        """
        Apply auto-transformation
        
        Args:
            context: Message context
            transformation: Transformation result from evaluation
            
        Returns:
            Transformation result
        """
        context_data = context.__dict__ if hasattr(context, '__dict__') else context
        transformation_data = transformation.__dict__ if hasattr(transformation, '__dict__') else transformation
        
        return self.client.request(
            "POST",
            f"{self.base_endpoint}/transform",
            json={
                "context": context_data,
                "transformation": transformation_data,
            }
        )
    
    def get_rules(self, tenant_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get all rules
        
        Args:
            tenant_id: Tenant ID
            
        Returns:
            List of rules
        """
        endpoint = f"{self.base_endpoint}/rules"
        if tenant_id:
            endpoint = f"{endpoint}/{tenant_id}"
        
        return self.client.request("GET", endpoint)
    
    def get_rule(
        self,
        rule_id: str,
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Get rule by ID
        
        Args:
            rule_id: Rule ID
            tenant_id: Tenant ID
            
        Returns:
            Rule dictionary
        """
        endpoint = f"{self.base_endpoint}/rules/{rule_id}"
        if tenant_id:
            endpoint = f"{self.base_endpoint}/rules/{tenant_id}/{rule_id}"
        
        return self.client.request("GET", endpoint)
    
    def create_rule(
        self,
        rule: Union[AutoTransformRule, Dict[str, Any]],
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create a new rule
        
        Args:
            rule: Rule to create
            tenant_id: Tenant ID
            
        Returns:
            Created rule information
        """
        rule_data = rule.__dict__ if hasattr(rule, '__dict__') else rule
        self._validate_rule(rule_data)
        
        endpoint = f"{self.base_endpoint}/rules"
        if tenant_id:
            endpoint = f"{endpoint}/{tenant_id}"
        
        return self.client.request("POST", endpoint, json=rule_data)
    
    def update_rule(
        self,
        rule_id: str,
        updates: Dict[str, Any],
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Update a rule
        
        Args:
            rule_id: Rule ID
            updates: Updates to apply
            tenant_id: Tenant ID
            
        Returns:
            Updated rule
        """
        endpoint = f"{self.base_endpoint}/rules/{rule_id}"
        if tenant_id:
            endpoint = f"{self.base_endpoint}/rules/{tenant_id}/{rule_id}"
        
        return self.client.request("PUT", endpoint, json=updates)
    
    def delete_rule(
        self,
        rule_id: str,
        tenant_id: Optional[str] = None,
    ) -> None:
        """
        Delete a rule
        
        Args:
            rule_id: Rule ID
            tenant_id: Tenant ID
        """
        endpoint = f"{self.base_endpoint}/rules/{rule_id}"
        if tenant_id:
            endpoint = f"{self.base_endpoint}/rules/{tenant_id}/{rule_id}"
        
        self.client.request("DELETE", endpoint)
    
    def enable_rule(
        self,
        rule_id: str,
        tenant_id: Optional[str] = None,
    ) -> None:
        """
        Enable a rule
        
        Args:
            rule_id: Rule ID
            tenant_id: Tenant ID
        """
        self.update_rule(rule_id, {"enabled": True}, tenant_id)
    
    def disable_rule(
        self,
        rule_id: str,
        tenant_id: Optional[str] = None,
    ) -> None:
        """
        Disable a rule
        
        Args:
            rule_id: Rule ID
            tenant_id: Tenant ID
        """
        self.update_rule(rule_id, {"enabled": False}, tenant_id)
    
    def get_templates(self) -> List[Dict[str, Any]]:
        """
        Get available templates
        
        Returns:
            List of templates
        """
        return self.client.request("GET", f"{self.base_endpoint}/templates")
    
    def apply_template(
        self,
        template_id: str,
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Apply a template
        
        Args:
            template_id: Template ID
            tenant_id: Tenant ID
            
        Returns:
            Applied rule information
        """
        endpoint = f"{self.base_endpoint}/apply-template/{template_id}"
        if tenant_id:
            endpoint = f"{self.base_endpoint}/apply-template/{tenant_id}/{template_id}"
        
        return self.client.request("POST", endpoint)
    
    def create_keyword_rule(
        self,
        name: str,
        keywords: List[str],
        transformation_type: Union[TransformationType, str],
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create keyword-based rule
        
        Args:
            name: Rule name
            keywords: Keywords to trigger on
            transformation_type: Transformation to apply
            tenant_id: Tenant ID
            
        Returns:
            Created rule information
        """
        rule = {
            "rule_name": name,
            "enabled": True,
            "priority": 0,
            "trigger_type": TriggerType.KEYWORD.value,
            "trigger_value": {"keywords": keywords},
            "transformation_type": transformation_type.value if isinstance(transformation_type, TransformationType) else transformation_type,
            "transformation_intensity": 2,
        }
        
        return self.create_rule(rule, tenant_id)
    
    def create_sentiment_rule(
        self,
        name: str,
        threshold: float,
        operator: str,
        transformation_type: Union[TransformationType, str],
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create sentiment-based rule
        
        Args:
            name: Rule name
            threshold: Sentiment threshold
            operator: Comparison operator
            transformation_type: Transformation to apply
            tenant_id: Tenant ID
            
        Returns:
            Created rule information
        """
        rule = {
            "rule_name": name,
            "enabled": True,
            "priority": 0,
            "trigger_type": TriggerType.SENTIMENT.value,
            "trigger_value": {"threshold": threshold, "operator": operator},
            "transformation_type": transformation_type.value if isinstance(transformation_type, TransformationType) else transformation_type,
            "transformation_intensity": 2,
        }
        
        return self.create_rule(rule, tenant_id)
    
    def create_time_rule(
        self,
        name: str,
        after: str,
        before: str,
        transformation_type: Union[TransformationType, str],
        tenant_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create time-based rule
        
        Args:
            name: Rule name
            after: Start time (HH:MM)
            before: End time (HH:MM)
            transformation_type: Transformation to apply
            tenant_id: Tenant ID
            
        Returns:
            Created rule information
        """
        rule = {
            "rule_name": name,
            "enabled": True,
            "priority": 0,
            "trigger_type": TriggerType.TIME.value,
            "trigger_value": {"after": after, "before": before},
            "transformation_type": transformation_type.value if isinstance(transformation_type, TransformationType) else transformation_type,
            "transformation_intensity": 2,
        }
        
        return self.create_rule(rule, tenant_id)
    
    def get_statistics(
        self,
        tenant_id: Optional[str] = None,
        period: int = 30,
    ) -> Dict[str, Any]:
        """
        Get rule statistics
        
        Args:
            tenant_id: Tenant ID
            period: Period in days
            
        Returns:
            Statistics dictionary
        """
        endpoint = f"{self.base_endpoint}/statistics"
        if tenant_id:
            endpoint = f"{endpoint}/{tenant_id}"
        
        return self.client.request("GET", endpoint, params={"period": period})
    
    def _validate_message_context(self, context: Dict[str, Any]) -> None:
        """Validate message context"""
        if not context.get("message"):
            raise ValidationError("Message is required")
        if not context.get("user_id"):
            raise ValidationError("User ID is required")
        if not context.get("tenant_id"):
            raise ValidationError("Tenant ID is required")
        if not context.get("platform"):
            raise ValidationError("Platform is required")
    
    def _validate_rule(self, rule: Dict[str, Any]) -> None:
        """Validate rule"""
        if not rule.get("rule_name"):
            raise ValidationError("Rule name is required")
        if not rule.get("trigger_type"):
            raise ValidationError("Trigger type is required")
        if not rule.get("trigger_value"):
            raise ValidationError("Trigger value is required")
        if not rule.get("transformation_type"):
            raise ValidationError("Transformation type is required")

File: sdk/javascript/src/services/auto-transform.ts
/**
 * Auto-Transform Service
 * Handles automatic transformation rules and configurations
 */

import { AxiosInstance } from 'axios';
import {
  AutoTransformConfig,
  AutoTransformRule,
  AutoTransformTemplate,
  MessageContext,
  TransformationResult,
  TriggerType
} from '../types';
import { ValidationError } from '../errors';

export class AutoTransformService {
  private baseUrl = '/auto-transform';

  constructor(private axios: AxiosInstance) {}

  /**
   * Get auto-transform configuration
   * @param tenantId Tenant ID
   */
  async getConfig(tenantId?: string): Promise<AutoTransformConfig> {
    const url = tenantId 
      ? `${this.baseUrl}/config/${tenantId}`
      : `${this.baseUrl}/config`;
    
    const response = await this.axios.get<AutoTransformConfig>(url);
    return response.data;
  }

  /**
   * Update auto-transform configuration
   * @param config Configuration to update
   * @param tenantId Tenant ID
   */
  async updateConfig(
    config: Partial<AutoTransformConfig>,
    tenantId?: string
  ): Promise<AutoTransformConfig> {
    const url = tenantId
      ? `${this.baseUrl}/config/${tenantId}`
      : `${this.baseUrl}/config`;

    const response = await this.axios.put<AutoTransformConfig>(url, config);
    return response.data;
  }

  /**
   * Enable auto-transform
   * @param tenantId Tenant ID
   */
  async enable(tenantId?: string): Promise<void> {
    await this.updateConfig({ enabled: true }, tenantId);
  }

  /**
   * Disable auto-transform
   * @param tenantId Tenant ID
   */
  async disable(tenantId?: string): Promise<void> {
    await this.updateConfig({ enabled: false }, tenantId);
  }

  /**
   * Evaluate if message should be auto-transformed
   * @param context Message context
   */
  async evaluate(context: MessageContext): Promise<TransformationResult> {
    this.validateMessageContext(context);

    const response = await this.axios.post<TransformationResult>(
      `${this.baseUrl}/evaluate`,
      context
    );

    return response.data;
  }

  /**
   * Apply auto-transformation
   * @param context Message context
   * @param transformation Transformation result from evaluation
   */
  async transform(
    context: MessageContext,
    transformation: TransformationResult
  ): Promise<{
    success: boolean;
    original: string;
    transformed: string;
    rule_applied: string;
    confidence: number;
  }> {
    const response = await this.axios.post(
      `${this.baseUrl}/transform`,
      {
        context,
        transformation
      }
    );

    return response.data;
  }

  /**
   * Get all rules
   * @param tenantId Tenant ID
   */
  async getRules(tenantId?: string): Promise<AutoTransformRule[]> {
    const url = tenantId
      ? `${this.baseUrl}/rules/${tenantId}`
      : `${this.baseUrl}/rules`;

    const response = await this.axios.get<AutoTransformRule[]>(url);
    return response.data;
  }

  /**
   * Get rule by ID
   * @param ruleId Rule ID
   * @param tenantId Tenant ID
   */
  async getRule(
    ruleId: string,
    tenantId?: string
  ): Promise<AutoTransformRule> {
    const url = tenantId
      ? `${this.baseUrl}/rules/${tenantId}/${ruleId}`
      : `${this.baseUrl}/rules/${ruleId}`;

    const response = await this.axios.get<AutoTransformRule>(url);
    return response.data;
  }

  /**
   * Create a new rule
   * @param rule Rule to create
   * @param tenantId Tenant ID
   */
  async createRule(
    rule: Omit<AutoTransformRule, 'id'>,
    tenantId?: string
  ): Promise<{ success: boolean; rule_id: string }> {
    this.validateRule(rule);

    const url = tenantId
      ? `${this.baseUrl}/rules/${tenantId}`
      : `${this.baseUrl}/rules`;

    const response = await this.axios.post(url, rule);
    return response.data;
  }

  /**
   * Update a rule
   * @param ruleId Rule ID
   * @param updates Updates to apply
   * @param tenantId Tenant ID
   */
  async updateRule(
    ruleId: string,
    updates: Partial<AutoTransformRule>,
    tenantId?: string
  ): Promise<AutoTransformRule> {
    const url = tenantId
      ? `${this.baseUrl}/rules/${tenantId}/${ruleId}`
      : `${this.baseUrl}/rules/${ruleId}`;

    const response = await this.axios.put<AutoTransformRule>(url, updates);
    return response.data;
  }

  /**
   * Delete a rule
   * @param ruleId Rule ID
   * @param tenantId Tenant ID
   */
  async deleteRule(ruleId: string, tenantId?: string): Promise<void> {
    const url = tenantId
      ? `${this.baseUrl}/rules/${tenantId}/${ruleId}`
      : `${this.baseUrl}/rules/${ruleId}`;

    await this.axios.delete(url);
  }

  /**
   * Enable a rule
   * @param ruleId Rule ID
   * @param tenantId Tenant ID
   */
  async enableRule(ruleId: string, tenantId?: string): Promise<void> {
    await this.updateRule(ruleId, { enabled: true }, tenantId);
  }

  /**
   * Disable a rule
   * @param ruleId Rule ID
   * @param tenantId Tenant ID
   */
  async disableRule(ruleId: string, tenantId?: string): Promise<void> {
    await this.updateRule(ruleId, { enabled: false }, tenantId);
  }

  /**
   * Get available templates
   */
  async getTemplates(): Promise<AutoTransformTemplate[]> {
    const response = await this.axios.get<AutoTransformTemplate[]>(
      `${this.baseUrl}/templates`
    );
    return response.data;
  }

  /**
   * Apply a template
   * @param templateId Template ID
   * @param tenantId Tenant ID
   */
  async applyTemplate(
    templateId: string,
    tenantId?: string
  ): Promise<{ success: boolean; rule_id: string }> {
    const url = tenantId
      ? `${this.baseUrl}/apply-template/${tenantId}/${templateId}`
      : `${this.baseUrl}/apply-template/${templateId}`;

    const response = await this.axios.post(url);
    return response.data;
  }

  /**
   * Create keyword-based rule
   * @param name Rule name
   * @param keywords Keywords to trigger on
   * @param transformationType Transformation to apply
   * @param tenantId Tenant ID
   */
  async createKeywordRule(
    name: string,
    keywords: string[],
    transformationType: string,
    tenantId?: string
  ): Promise<{ success: boolean; rule_id: string }> {
    return this.createRule(
      {
        rule_name: name,
        enabled: true,
        priority: 0,
        trigger_type: 'keyword',
        trigger_value: { keywords },
        transformation_type: transformationType as any,
        transformation_intensity: 2
      },
      tenantId
    );
  }

  /**
   * Create sentiment-based rule
   * @param name Rule name
   * @param threshold Sentiment threshold
   * @param operator Comparison operator
   * @param transformationType Transformation to apply
   * @param tenantId Tenant ID
   */
  async createSentimentRule(
    name: string,
    threshold: number,
    operator: 'less_than' | 'greater_than' | 'equals',
    transformationType: string,
    tenantId?: string
  ): Promise<{ success: boolean; rule_id: string }> {
    return this.createRule(
      {
        rule_name: name,
        enabled: true,
        priority: 0,
        trigger_type: 'sentiment',
        trigger_value: { threshold, operator },
        transformation_type: transformationType as any,
        transformation_intensity: 2
      },
      tenantId
    );
  }

  /**
   * Create time-based rule
   * @param name Rule name
   * @param after Start time (HH:MM)
   * @param before End time (HH:MM)
   * @param transformationType Transformation to apply
   * @param tenantId Tenant ID
   */
  async createTimeRule(
    name: string,
    after: string,
    before: string,
    transformationType: string,
    tenantId?: string
  ): Promise<{ success: boolean; rule_id: string }> {
    return this.createRule(
      {
        rule_name: name,
        enabled: true,
        priority: 0,
        trigger_type: 'time',
        trigger_value: { after, before },
        transformation_type: transformationType as any,
        transformation_intensity: 2
      },
      tenantId
    );
  }

  /**
   * Get rule statistics
   * @param tenantId Tenant ID
   * @param period Period in days
   */
  async getRuleStatistics(
    tenantId?: string,
    period: number = 30
  ): Promise<{
    total_evaluations: number;
    total_transformations: number;
    rule_usage: Record<string, number>;
    success_rate: number;
  }> {
    const url = tenantId
      ? `${this.baseUrl}/statistics/${tenantId}`
      : `${this.baseUrl}/statistics`;

    const response = await this.axios.get(url, {
      params: { period }
    });

    return response.data;
  }

  /**
   * Validate message context
   */
  private validateMessageContext(context: MessageContext): void {
    if (!context.message) {
      throw new ValidationError('Message is required');
    }
    if (!context.user_id) {
      throw new ValidationError('User ID is required');
    }
    if (!context.tenant_id) {
      throw new ValidationError('Tenant ID is required');
    }
    if (!context.platform) {
      throw new ValidationError('Platform is required');
    }
  }

  /**
   * Validate rule
   */
  private validateRule(rule: Omit<AutoTransformRule, 'id'>): void {
    if (!rule.rule_name) {
      throw new ValidationError('Rule name is required');
    }
    if (!rule.trigger_type) {
      throw new ValidationError('Trigger type is required');
    }
    if (!rule.trigger_value) {
      throw new ValidationError('Trigger value is required');
    }
    if (!rule.transformation_type) {
      throw new ValidationError('Transformation type is required');
    }
  }
}